### PR TITLE
Enable VP9 SVC with webrtc m94

### PR DESCRIPTION
**Pull request by harvestsure**
Enable VP9 SVC with webrtc m94 webrtc::RtpEncodingParameters.scalability_mode

### CODE REVIEW

File: src/Handler.cpp [hunk 1]
```diff
@@ -194,23 +208,45 @@ namespace mediasoupclient
...
+		// Special case for VP9 with SVC.
+		bool hackVp9Svc = false;
 
 		try
 		{
 			webrtc::PeerConnectionInterface::RTCOfferAnswerOptions options;
 
 			offer               = this->pc->CreateOffer(options);
-			auto localSdpObject = sdptransform::parse(offer);
+			json localSdpObject = sdptransform::parse(offer);
```
Comment (ibc) [inline src/Handler.cpp]: Why this change?

**Review by ibc — COMMENTED** [review]
I'll continue review later

File: src/Handler.cpp [hunk 2]
```diff
@@ -224,7 +260,7 @@ namespace mediasoupclient
 		catch (std::exception& error)
 		{
 			// Panic here. Try to undo things.
-			transceiver->SetDirection(webrtc::RtpTransceiverDirection::kInactive);
+			//transceiver->SetDirection(webrtc::RtpTransceiverDirection::kInactive);
```
Comment (ibc) [inline src/Handler.cpp]: Why is this commented? If it doesn't make sense let's just remove the line. But why?

File: src/Handler.cpp (hunk 1 above)
Comment (harvestsure) [reply src/Handler.cpp]: Nothing, just my personal habit. You can ignore it

File: src/mediasoupclient.cpp [hunk 3]
```diff
@@ -16,6 +17,8 @@ namespace mediasoupclient
 
 		MSC_DEBUG("mediasoupclient v%s", Version().c_str());
 
+		webrtc::field_trial::InitFieldTrialsFromString("WebRTC-SupportVP9SVC/EnabledByFlag_3SL3TL/");
```
Comment (ibc) [inline src/mediasoupclient.cpp:20]: I'm not sure this is the proper place to set those libwebrtc flags (maybe). @jmillan?

File: src/sdp/Utils.cpp [hunk 4]
```diff
@@ -291,7 +291,7 @@ namespace mediasoupclient
 						return false;
 
 					return (
-					  jsonAttributeIt->get<std::string>() == "cname" && jsonIdIt->get<uint32_t>() == firstSsrc);
+					  jsonAttributeIt->get<std::string>() == "cname" /*&& jsonIdIt->get<uint32_t>() == firstSsrc*/);
```
Comment (ibc) [inline src/sdp/Utils.cpp]: Why this?

File: src/sdp/Utils.cpp [hunk 5]
```diff
@@ -390,7 +390,7 @@ namespace mediasoupclient
 			{
 				MSC_TRACE();
 
-				auto jsonMssrcsIt = offerMediaObject.find("ssrc");
```
Comment (ibc) [inline src/sdp/Utils.cpp:393]: Why this?

File: src/Handler.cpp (hunk 1 above)
Comment (ibc) [reply src/Handler.cpp]: I cannot ignore it, it's part of your proposed PR 😀

File: src/Handler.cpp (hunk 2 above)
Comment (harvestsure) [reply src/Handler.cpp]: just mark it,maybe i will change lower version of webrtc

File: src/sdp/Utils.cpp (hunk 4 above)
Comment (harvestsure) [reply src/sdp/Utils.cpp]: I refer to project mediasoup-client, 
`const ssrcCnameLine = offerMediaObject.ssrcs
		.find((line: any) => line.attribute === 'cname');

	if (!ssrcCnameLine)
		throw new Error('a=ssrc line with cname information not found');`

File: src/sdp/Utils.cpp (hunk 5 above)
Comment (harvestsure) [reply src/sdp/Utils.cpp:393]: I refer to project mediasoup-client,
`export function getCname(
	{ offerMediaObject }:
	{ offerMediaObject: any }
): string
{
	const ssrcCnameLine = (offerMediaObject.ssrcs || [])
		.find((line: { attribute: string }) => line.attribute === 'cname');

	if (!ssrcCnameLine)
		return '';

	return ssrcCnameLine.value;
}`

File: src/Handler.cpp (hunk 1 above)
Comment (harvestsure) [reply src/Handler.cpp]: I'll pay attention next time 🙂

Comment (jmillan) [reply src/Handler.cpp]: yes please, move it back to `auto`.

File: src/mediasoupclient.cpp (hunk 3 above)
Comment (jmillan) [reply src/mediasoupclient.cpp:20]: Yes, this can be placed here.

File: src/sdp/Utils.cpp (hunk 4 above)
Comment (ibc) [reply src/sdp/Utils.cpp]: Ok, then please remove the commented code. @jmillan do you know why we were doing this different here than in mediasoup-client? It doesn't make any sense to me.

File: src/sdp/Utils.cpp (hunk 5 above)
Comment (ibc) [reply src/sdp/Utils.cpp:393]: ok

File: src/Handler.cpp (hunk 2 above)
Comment (ibc) [reply src/Handler.cpp]: I don't understand. How does lower version of webrtc affect here? My question is: why is that line commented in your PR?

File: src/Handler.cpp (hunk 1 above)
Comment (harvestsure) [reply src/Handler.cpp]: I have modified it

File: src/Handler.cpp (hunk 2 above)
Comment (harvestsure) [reply src/Handler.cpp]: I'm careless. The old interface has been deprecated：
RTC_DEPRECATED virtual void SetDirection(
      RtpTransceiverDirection new_direction);

File: src/sdp/Utils.cpp (hunk 4 above)
Comment (harvestsure) [reply src/sdp/Utils.cpp]: removed🙂

**Review by ibc — APPROVED** [review]
Nice

**Comment by ibc** [comment]
@jmillan anything else from your side? should we merge and release?

**Comment by ibc** [comment]
@harvestsure I'd appreciate a PR in the mediasoup-website project with updated API documentation.

**Comment by jmillan** [comment]
> @jmillan anything else from your side? should we merge and release?

It looks good to me. I'Il merge it and release tomorrow if you don't do it before.

**Comment by ibc** [comment]
Sorry for the delay, terribly busy. Will check this in next days but cannot provide with a proper ETA yet.

**Comment by ibc** [comment]
@jmillan ping

**Comment by jmillan** [comment]
I'm testing this PR with m89.

**Comment by jmillan** [comment]
Actually m89 fails to compile with Xcode 13 and it's already very old. I'll try the newest stable branch instead.

**Comment by jmillan** [comment]
@harvestsure,

You didn't modify the tests, which do not compile. Please, adapt the test files accordingly and also, let me ask you to target this PR against the new libmediasoupclient which uses M94.

**Comment by harvestsure** [comment]
> @harvestsure,
> 
> You didn't modify the tests, which do not compile. Please, adapt the test files accordingly and also, let me ask you to target this PR against the new libmediasoupclient which uses M94.

I have compiled using the M94 version with test files.

**Comment by jmillan** [comment]
There's a warning here @harvestsure,


Screenshot 2021-11-29 at 13 38 08

Also, since this changes the API we'd  appreciate a PR in the mediasoup-website project with updated API documentation

**Comment by harvestsure** [comment]
> There's a warning here @harvestsure,
> 
> Screenshot 2021-11-29 at 13 38 08
> 
> Also, since this changes the API we'd appreciate a PR in the mediasoup-website project with updated API documentation

I'll deal with it later

**Comment by jmillan** [comment]
Thanks, the doc will be updated.
